src/Type.hs

```diff
@@ -214,30 +212,23 @@ checkContract store constructors (U.Contract constr@(U.Definition _ cid _ _ _ _
 -- checks a transition given a typing of its storage variables
 checkTransition :: Env -> U.Transition -> Err [Behaviour]
 checkTransition env (U.Transition _ name contract iface@(Interface _ decls) iffs cases posts) =
-  noIllegalWilds *>
   -- constrain integer calldata variables (TODO: other types)
   fmap fmap (makeBehv <$> checkIffs env' iffs <*> traverse (checkExpr env' SBoolean) posts)
-  <*> traverse (checkCase env') normalizedCases
+  <*> traverse (checkCase env') elaborateCases
   where
     env' = addCalldata env decls
 
-    noIllegalWilds :: Err ()
-    noIllegalWilds = case cases of
-      U.Direct   _  -> pure ()
-      U.Branches bs -> for_ (init bs) $ \c@(U.Case p _ _) ->
-                          when (isWild c) (throw (p, "Wildcard pattern must be last case"))  -- TODO test when wildcard isn't last
-
-    -- translate wildcards into negation of other branches and translate a single case to a wildcard
-    normalizedCases :: [U.Case]
-    normalizedCases = case cases of
-      U.Direct   post -> [U.Case nowhere (U.WildExp nowhere) post]
-      U.Branches bs ->
-        let
-          Just (rest, lastCase@(U.Case pn _ post)) = unsnoc bs
-          negation = U.ENot nowhere $
-                        foldl (\acc (U.Case _ e _) -> U.EOr nowhere e acc) (U.BoolLit nowhere False) rest
-        in rest `snoc` (if isWild lastCase then U.Case pn negation post else lastCase)
-    -- TODO ensure non-overlapping and exhaustiveness (maybe with elaboration and mandatory wildcard?) 
+    -- make cases non overlapping by negating previous cases that haven't matched
+    elaborateCases :: [U.Case]
+    elaborateCases = case cases of
+      U.Direct   post -> [U.Case nowhere (U.BoolLit nowhere True) post]
+      U.Branches [] -> error "Internal error: branches cannot be empty"
+      U.Branches (U.Case p cond post:cases') -> (U.Case p cond post):makeCases (U.ENot nowhere cond) cases'
+
+    makeCases :: U.Expr -> [U.Case] -> [U.Case]
+    makeCases _ [] = []
+    makeCases neg (U.Case p cond post:cases') = (U.Case p (U.EAnd nowhere cond neg) post):(makeCases (U.EAnd nowhere (U.ENot nowhere cond) neg) cases')
+    -- TODO ensure that branches are exhaustive
```

Comment:
What would be the storage and return value in case of an implicit wildcard case? 

One easy way, which is rather unsatisfying syntactically, is to enforce the last case to be a wildcard. 

A harder one, is to ask the SMT solver, during typechecking, whether the cases are exhaustive.